docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,205 @@
+
+#### KubeAPIServerUnreachableInternally
+
+The `KubeAPIServerUnreachableInternally` condition type can be implemented (both the monitoring and the remedial action parts) similar to the [`ETCDUnreachable`](#etcdunreachable) condition type.
+
+#### UnhealthyNodesThresholdReached
+
+A component can monitor the fraction of the `NotReady` nodes in a zone and update the conditions with the `UnhealthyNodesThresholdReached` condition type.
+The list of affected zones could be maintained in the [`Message`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L168).
+
+##### Possible Actions
+
+- Alerts can be raised to the administrator of the cluster as well as the operator/support team on duty if this condition lasts for some duration.
+- If necessary, `kube-controller-manager` could be scaled down as described [above](#possible-actions).
+- Some resource creations (say, `pods`) could be throttled/rate-limited.
+- The `machine-controller-manager` could be either scaled down as described [above](#possible-actions) or be enhanced to avoid (or ramp down) deletion of machines from the affected zones (or `machinedeployments`).
+- Gardener can defer any disruptive updates until later.
+
+### Condition Types of future relevance
+
+This is just a sample set of condition types that might be of future relevance. It is not an exhaustive list.
+The suggested possible actions for these condition types are also just samples and not an exhaustive list.
+
+#### UnhealthyInternalNetwork
+
+If there is a way to detect problems in the cluster network that is severe enough to affect any significant portion of the cluster then this could be encoded in a component that updates the conditions with the `UnhealthyInternalNetwork` condition type.
+
+##### Possible Actions
+
+- Alerts can be raised to the administrator of the cluster as well as the operator/support team on duty.
+- Some resource creations (say, `pods`) could be throttled/rate-limited.
+- Gardener can defer any disruptive updates until later.
+
+#### UnhealthyCloudProvider
+
+If there is information about some issues reported against the cloud provider, it could be updated in the conditions with the `UnhealthyCloudProvider` condition type.
+If there is more specific information about the issue then more specific condition types can be defined.
+
+Such conditions could be updated manually by the administrators, operators on duty based on publications of such issues by the cloud providers.
```

Comment:
I cannot image how the conditions for thousands of cluster are updated manually.
This is error prone process, and it can be ensured that the conditions will contain the most actual status.

Reply:
> I cannot imagine how the conditions for thousands of cluster are updated manually.

I didn't mean manually for every one of the `clusters`. The point here is that the source of some conditions could be ad hoc and not necessarily automatically monitored.

Once there is a `Cluster` resource that aggregates the health conditions related to a particular cluster, it is would be possible to build on it. That is the whole point of the operator concept. We could introduce a resource to aggregate health conditions for a cloud provider (perhaps even the existing `CloudProfile`) from which the relevant `status.conditions` could be replicated automatically to the corresponding `clusters`